analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Collections.cs

```diff
@@ -0,0 +1,76 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors;
+
+internal sealed partial class Binary
+{
+    private ProgramState LearnBranchingCollectionConstraint(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
+    {
+        var collection = InstanceOfCountPropertyReference(binary.LeftOperand);
+        var countIsLeftOperand = true;
+        if (collection is null)
+        {
+            collection = InstanceOfCountPropertyReference(binary.RightOperand);
+            countIsLeftOperand = false;
+        }
+
+        return collection is not null
+            && state.Constraint<NumberConstraint>(OtherOperator(binary, countIsLeftOperand)) is { } number
+            && BranchingCollectionConstraint(binary.OperatorKind, falseBranch, countIsLeftOperand, number) is { } constraint
+                ? state.SetSymbolConstraint(collection, constraint)
+                : state;
```

Comment:
I find this `countIsLeftOperand ` very confusing.
What if you take this logic and put in a method where you give the binary operation as input and you return a bool if it's needed to be flipped or not?

Reply:
Assume that the count is on the left
```csharp 
bool ShouldFlip(Operation binary) 
{
     //Check if the property count is actually left
    //If yes return true
}

var shouldFlip = ShouldFlip(arg);

var count = shouldFlip ? left: right
var otherSideOfCount = shouldFlip? right : left
var operator = shouldFlip? asIs : opposite
```
Based on `ShouldFlip` you take all your decisions.
It's what you have already right now but to my eyes a bit more readable.